File: libraries/javalib/java/lang/Throwable.java
/*
 * Java core library component.
 *
 * Copyright (c) 1997, 1998
 *      Transvirtual Technologies, Inc.  All rights reserved.
 *
 * See the file "license.terms" for information on usage and redistribution
 * of this file.
 */

package java.lang;

import java.io.PrintStream;
import java.io.PrintWriter;

public class Throwable extends Object
{
	private Object backtrace = null;
	private String message = null;

public Throwable()
	{
	message = null;
	fillInStackTrace();
}

public Throwable(String mess)
	{
	message = mess;
	fillInStackTrace();
}

native public Throwable fillInStackTrace();

public String getLocalizedMessage()
	{
	return (getMessage());
}

public String getMessage()
	{
	return (message);
}

public void printStackTrace()
	{
	printStackTrace(System.err);
}

public void printStackTrace(PrintStream s)
	{
	if (message != null) {
		s.println(this.getClass().getName() + ": " + message);
	}
	else {
		s.println(this.getClass().getName());
	}
	printStackTrace0(s);
}

public void printStackTrace(PrintWriter s)
	{
	if (message != null) {
		s.println(this.getClass().getName() + ": " + message);
	}
	else {
		s.println(this.getClass().getName());
	}
	printStackTrace0(s);
}

native private void printStackTrace0(Object s);

public String toString()
	{
	return (this.getClass().getName() + ": " + message);
}
}

File: libraries/clib/native/Throwable.c
/*
 * java.lang.Throwable.c
 *
 * Copyright (c) 1996, 1997
 *	Transvirtual Technologies, Inc.  All rights reserved.
 * Copyright (c) 2003
 *	Mark J. Wielaard <mark@klomp.org>
 *
 * See the file "license.terms" for information on usage and redistribution 
 * of this file. 
 */

#include "config.h"
#include "config-io.h"
#include <assert.h>
#include <native.h>
#include "java_lang_StackTraceElement.h"
#include "java_lang_Throwable.h"
#include "java_lang_VMThrowable.h"

extern Hjava_lang_Object* buildStackTrace(void*);
extern HArrayOfObject* getStackTraceElements(struct Hjava_lang_VMThrowable*,
					     struct Hjava_lang_Throwable*);

/*
 * Fill in stack trace information - don't know what thought.
 */
void
java_lang_VMThrowable_fillInStackTrace(struct Hjava_lang_VMThrowable* o)
{
	unhand(o)->vmdata = buildStackTrace(NULL);
	assert(unhand(o)->vmdata != NULL);
}

/*
 * Returns a array of StackTraceElements for the given VMThrowable state.
 */
HArrayOfObject*
java_lang_VMThrowable_getStackTrace(struct Hjava_lang_VMThrowable* state, struct Hjava_lang_Throwable* throwable)
{
	return getStackTraceElements(state, throwable);
}
